### PR TITLE
feat(ui): allow selecting model in answer engine

### PR DESCRIPTION
**Pull request by antimonyGu**
front end of  #3097 

- When `supported_model` is not set, 
no-supported-model

- When  `supported_model` set,
[recording](https://jam.dev/c/b1861c39-85b4-465d-8661-c17b5e9ed8dd)

### CODE REVIEW

File: ee/tabby-ui/components/textarea-search.tsx [hunk 1]
```diff
@@ -53,12 +68,45 @@ export default function TextAreaSearch({
...
+            .map(key => modelInfo[key as keyof typeof modelInfo])
+            .flat()
+            .filter(val => val)
+        : [],
+    [modelInfo]
+  )
+  const [selectedModel, setSelectedModel] = useState(
+    modelInfoArray.length > 0 ? modelInfoArray[0] : ''
+  )
+  const isSelecteEnabled = false
```
Comment (antimonyGu) [inline ee/tabby-ui/components/textarea-search.tsx:85]: `isSelecteEnabled` as a feature toggle when BE is ready will flip this toggle.

File: ee/tabby-schema/src/schema/thread/inputs.rs [hunk 2]
```diff
@@ -70,6 +70,9 @@ fn validate_code_query_input(input: &CodeQueryInput) -> Result<(), ValidationErr
 
 #[derive(GraphQLInputObject, Validate, Default, Clone)]
 pub struct ThreadRunOptionsInput {
+    #[graphql(default)]
+    pub selected_model_name: Option<String>,
```
Comment (wsxiaoys) [inline ee/tabby-schema/src/schema/thread/inputs.rs:74]:
```suggestion
    pub model_name: Option<String>,
```

File: ee/tabby-webserver/src/service/answer.rs [hunk 3]
```diff
@@ -148,17 +148,22 @@ impl AnswerService {
...
 
+            println!("selected_model_name: {:?}", options.selected_model_name);
+
             // 4. Prepare requesting LLM
             let request = {
                 let chat_messages = convert_messages_to_chat_completion_request(&self.config, &context_info_helper, &messages, &attachment, user_attachment_input.as_ref())?;
 
                 CreateChatCompletionRequestArgs::default()
                     .messages(chat_messages)
+                    .model(options.selected_model_name.as_deref().unwrap_or(""))
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/answer.rs:159]: only call function when model name is not None

File: ee/tabby-ui/app/(home)/page.tsx [hunk 4]
```diff
@@ -46,6 +47,10 @@ function MainPanel() {
     query: contextInfoQuery
   })
   const scrollY = useStore(useScrollStore, state => state.homePage)
+  const [selectedModel, setSelectedModel] = useLocalStorage<string>(
```
Comment (wsxiaoys) [inline ee/tabby-ui/app/(home)/page.tsx:50]: please handle situation if current persisted selectedModel doesn't exist in `/v1beta/models`

File: crates/tabby-inference/src/chat.rs [hunk 5]
```diff
@@ -54,7 +54,9 @@ impl ExtendedOpenAIConfig {
         &self,
         mut request: CreateChatCompletionRequest,
     ) -> CreateChatCompletionRequest {
-        request.model = self.model_name.clone();
+        if request.model.is_empty() {
+            request.model = self.model_name.clone();
```
Comment (wsxiaoys) [inline crates/tabby-inference/src/chat.rs:58]: maybe good to check request.model has to belong to `supported_models` here (if not, generate a warning and set request.model = self.model_name)

File: ee/tabby-ui/components/textarea-search.tsx [hunk 6]
```diff
@@ -199,57 +250,79 @@ export default function TextAreaSearch({
       </div>
       <div
```
Comment (liangfung) [inline ee/tabby-ui/components/textarea-search.tsx:251]: If the value of modelInfo.chat is null, consider not displaying the bottom toolbar.

File: ee/tabby-ui/components/textarea-search.tsx [hunk 7]
```diff
@@ -54,12 +80,34 @@ export default function TextAreaSearch({
   const [value, setValue] = useState('')
   const { theme } = useCurrentTheme()
   const editorRef = useRef<PromptEditorRef>(null)
+  const { data: modelInfo } = useModel()
+  const isSelectModelEnabled = true
+
+  const DropdownMenuItems = modelInfo?.chat.map(model => (
```
Comment (liangfung) [inline ee/tabby-ui/components/textarea-search.tsx:86]: The value of `modelInfo?.chat` may be null, which can cause an error.

**Review by liangfung — COMMENTED** [review]
The function `onRegenerateResponse` requires the `modelName` parameter to be passed in.

File: crates/tabby-inference/src/chat.rs (hunk 5 above)
Comment (antimonyGu) [reply crates/tabby-inference/src/chat.rs:58]: needs add field `supported_models` to ExtendedOpenAIConfig to

File: crates/tabby-inference/src/chat.rs [hunk 8]
```diff
@@ -56,6 +60,22 @@ impl ExtendedOpenAIConfig {
...
+                    HeaderFormat::BoldRed,
+                    &[
+                        &format!(
+                            "Warning: {} Model is not supported, falling back to {}",
+                            request.model,
+                            self.model_name
+                        )
+                    ],
+                )
+                .print();
```
Comment (wsxiaoys) [inline crates/tabby-inference/src/chat.rs:76]: use `warn!` to generate a warining.

**Pull request merged**